src/modifierThorns.ts

```diff
@@ -0,0 +1,48 @@
+import { registerEvents, registerModifiers } from "./cards";
+import { getOrInitModifier } from "./cards/util";
+import * as Unit from './entity/Unit';
+import Underworld from './Underworld';
+
+// Deals (quantity) damage to an attacker when taking damage
+export const thornsId = 'Thorns';
+export default function registerThorns() {
+  registerModifiers(thornsId, {
+    description: 'Deals (quantity) damage to an attacker when taking damage',
+    add: (unit: Unit.IUnit, underworld: Underworld, prediction: boolean, quantity: number = 1) => {
+      getOrInitModifier(unit, thornsId, { isCurse: false, quantity, keepOnDeath: true }, () => {
+        Unit.addEvent(unit, thornsId);
+      });
+
+      if (!prediction) {
+        updateTooltip(unit);
+      }
+    }
+  });
+  registerEvents(thornsId, {
+    onTakeDamage: (unit: Unit.IUnit, amount: number, underworld: Underworld, prediction: boolean, damageDealer?: Unit.IUnit) => {
+      const modifier = unit.modifiers[thornsId];
+      if (modifier) {
+        // Thorns will not deal damage if we are being healed
+        if (damageDealer && amount > 0) {
+          // Deal flat damage to the attacker
+          Unit.takeDamage({
+            unit: damageDealer,
+            amount: modifier.quantity,
+            sourceUnit: unit,
+          }, underworld, prediction);
```

Comment:
Maybe we add an optional param to skip the onTakeDamage event so that we can avoid a thorns loop here?

Reply:
Maybe, but that would also allow thorns to bypass fortify, armor, invulnerability, debilitate, etc. I think our answer has to be more nuanced than that for some effects.

- We could differentiate between "on hit" and "on damage" events, such that onDamage events don't proc reactive onHit effects like thorns, but still check for the non-reactive effects like armor? We may still need to add a couple optional parameters on top of that as we go, like isDamageModifiable or something.
- Some games will keep track of what reactive effects have triggered from one "chain" of damage events and prevent them from triggering more than once.

When we find a modular solution, we should use it to replace the old Soul Bind logic as well.

Reply:
Maybe "isHit" could be a boolean/optional parameter in the onTakeDamage event

Reply:
Hmm, I don't want to over complicate it but I see what you mean.  Rather than adding extra events, what if we have a greater takeDamage callback that triggers everything (events and all) and then a secret inner function that just handles non-reactive stuff that is named such that it is obviously not meant to be used outside of special occasions.  Then SoulBind and Thorns could use the inner one